packages/gatsby/src/commands/serve.js

```diff
@@ -85,9 +85,20 @@ module.exports = async program => {
   app.use(telemetry.expressMiddleware(`SERVE`))
 
   router.use(compression())
-  router.use(express.static(`public`))
+  router.use(express.static(`public`, { redirect: false }))
   const matchPaths = await readMatchPaths(program)
   router.use(matchPathRouter(matchPaths, { root }))
+  router.use((req, res, next) => {
+    // Handle URLs without trailing slashes, e.g. /hello => /hello/index.html
+    // Express already handles trailing slashes, e.g. /hello/ => /hello/index.html
+    if (req.url.slice(-1) !== `/`) {
+      res.sendFile(req.url + `/index.html`, {
+        root,
+      })
```

Comment:
What if the file doesn't exist? A callback argument can be added here to deal with this case and allow for the regular 404 logic to occur:

```
res.sendFile(...., (error) => {
  if (error) next()
  // if there's no error, request.end() has been called and there's nothing to do.
})
```